### DISCUSSION TITLE
insertmanyvalues issue with custom type

### DISCUSSION THREAD

**Discussion by al-dpopowich** [sqlalchemy/sqlalchemy#10483] [Usage Questions] [answered]
A number of issues were opened due to work on #9618 , regressions when inserting multiple rows with columns of a certain type (e.g., UUID - #9808, LargeBinary - #9739).

In my case I have a custom type: I subclass `sqlalchemy.dialects.postgresql.Range`,  which in turn wraps `asyncpg.Range`.  When I flush an AsyncSession with more than one new ORM instance, I get a low-level asyncpg exception:

```
(sqlalchemy.dialects.postgresql.asyncpg.Error) <class 'asyncpg.exceptions.DataError'>: invalid input for query argument $4: <Range [2023, 2023]> (expected str, got Range)
[SQL: INSERT INTO questionnaire (company_id, study_id, project_id, years, title, completed, user_id) SELECT p0::INTEGER, p1::INTEGER, p2::INTEGER, p3::TYRANGE, p4::VARCHAR, p5::TIMESTAMP WITH TIME ZONE, p6::VARCHAR FROM (VALUES ($1::INTEGER, $2::INTEGER, $ ... 322 characters truncated ...  p4, p5, p6, sen_counter) ORDER BY sen_counter RETURNING questionnaire.id, questionnaire.id AS id__1]
[parameters: (1, 12, None, <Range [2023, 2023]>, 'Profile Questionnaire for customer@dev.xyz', None, 'b36bd7f9-d779-49ce-a4be-1864592cbcb5', 1, 12, None, <Range [2023, 2023]>, 'Profile Questionnaire for jdoe@internal.local', None, 'c5e1693f-74d5-43b3-b19a-1fe95998baa1', 1, 12, None, <Range [2023, 2023]>, 'Profile Questionnaire for janed@internal.local', None, '826ce3f4-7a7e-42ec-b71a-d04485320c3c')]
(Background on this error at: https://sqlalche.me/e/20/dbapi)
```

This worked in `2.0.9`, but breaks in `2.0.10+`

I've looked at the fixes for UUID and LargeBinary, etc, but it's not immediately obvious to me what I need to do to fix this for a custom type.

Thanks!

**Comment by zzzeek** [comment]
hi - 

I need an MCVE please, work on this issue is ongoing as long as we keep getting regressions, thanks!

**Reply by al-dpopowich** [reply]
I was afraid of that.  ;-)

Might take me a while for the MCVE, given the complexity of the custom type.

**Reply by zzzeek** [reply]
if you are subclassing Range are you also using a TypeDecorator for type conversion?  What's the DB type in use?

**Reply by al-dpopowich** [reply]
It gets complicated...

In postgres, I have a custom range type:

```sql
CREATE TYPE tyrange AS RANGE (
   subtype = taxyear
)
```

In turn, `taxyear` is a custom `DOMAIN`, essentially, an integer with a check constraint for years that are logical for our application.  Both `taxyear` and `tyrange` save a ton of validation code in my application.

In sqla, I have:

```python
class TYRANGE(UserDefinedType):
   """SQLA column type for custom type TYRANGE supported in python by
   class TaxYearRange
   """

   python_type = TaxYearRange  # <-- this is a python class I use in app code, a subclass of sqlalchemy.dialects.postgresql.Range

   cache_ok = True

   comparator_factory = AbstractRange.comparator_factory

   def get_col_spec(self): # pragma: no cover
      """PG type"""
      return "TYRANGE"

   def bind_processor(self, dialect):
      """Convert TaxYearRange TYRANGE form for pg.
      """
      # details left out for brevity

   def result_processor(self, _dialect, _coltype):
      """Convert Range as returned from asyncpg to TaxYearRange
      """
      # details left out for brevity
```

This worked like a charm in `2.0.9` on down (first used in `1.4.x`).  Breaks in `2.0.10+`.

**Reply by zzzeek** [reply, marked as answer]
~~we support custom domains~~. never mind that, we dont support the composed type part you have there

UserDefinedType should have some path to working, which here would be adding some combination of:

```
    render_bind_cast = True
    render_literal_cast = True
```

to your custom type.  all those regressions were because the PG-specific types which have those flags were not being used, but in this case, your UserDefinedType is it, so you'd set those flags to control the rendering behaviors.

**Reply by al-dpopowich** [reply]
Setting the `render_bind_cast` attribute did it!

I'm curious, is there a reason these are not `True` by default?  When wouldn't you want them?

**Reply by zzzeek** [reply]
well they create a lot of verbosity and noise in the SQL but also there's a chance that they get in the way of the driver's own ability to infer types in some cases.

**Reply by zzzeek** [reply]
they should likely be more prominently documented for UserDefinedType though.

**Reply by al-dpopowich** [reply]
Thanks, as always, for your prompt, on-target replies.

(Do you ever sleep?)  ;-)

**Reply by zzzeek** [reply]
well it's work hours right now, 9-12 ET is typically most of my emailing time
